src/utils/decodedSourcemap.ts

```diff
@@ -22,7 +67,37 @@ export function decodedSourcemap(map: Input): ExistingDecodedSourceMap | null {
 		};
 	}
 
-	const mappings = typeof map.mappings === 'string' ? decode(map.mappings) : map.mappings;
+	const originalMappings = map.mappings;
+	const isAlreadyDecoded = typeof originalMappings !== 'string';
+	const cache = {
+		decodedMappings: isAlreadyDecoded ? originalMappings : undefined,
+		encodedMappings: isAlreadyDecoded ? undefined : originalMappings
+	};
+
+	const decodedMap = {
+		...(map as ExistingRawSourceMap | ExistingDecodedSourceMap),
+		// By moving mappings behind an accessor, we can avoid unneeded computation for cases
+		// where the mappings field is never actually accessed. This appears to greatly reduce
+		// the overhead of sourcemap decoding in terms of both compute time and memory usage.
+		get mappings() {
+			if (cache.decodedMappings) {
+				return cache.decodedMappings;
+			}
+			// If decodedMappings doesn't exist then encodedMappings should.
+			// The only scenario where cache.encodedMappings should be undefined is if the map
+			// this was constructed from was already decoded, or if mappings was set to a new
+			// decoded string. In either case, this line shouldn't get hit.
+			cache.decodedMappings = decode(cache.encodedMappings!);
+			cache.encodedMappings = undefined;
+			return cache.decodedMappings;
+		},
+		set mappings(value) {
```

Comment:
Do we need the setter? Coverage is complaining that it is not used.

Reply:
I'm happy to drop it. I would expect the cases where someone might be modifying mappings on an existing sourcemap to be rare. I think they would need to hook into the cache and modify it there. That would now fail because the property is not writeable, but the workaround for them is to just replace the entire sourcemap. It seems like a pretty fringe edge case though, so avoiding the setter sounds good to me.